### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [5]:
# Dependencies and Setup
import pandas as pd
import os
# File to Load (Remember to Change These)
data_file = os.path.join("Resources","purchase_data.csv")

# Read Purchasing File and store into Pandas data frame
data_file_df = pd.read_csv(data_file)
#look at the first 5 to see what data we have
data_file_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [6]:
#valuecount will give unquie users and len will give us the numbers of them
num_of_users = data_file_df["SN"].value_counts()
Total_users = len(num_of_users)
Total_users

576

In [7]:
print("Total number of players ", Total_users)

Total number of players  576


## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [8]:
#getting rid of nas if any and clean data

data_file_df.dropna(how='any')

Purchase ID             SN  Age  Gender  Item ID  \
0              0        Lisim78   20    Male      108   
1              1    Lisovynya38   40    Male      143   
2              2     Ithergue48   24    Male       92   
3              3  Chamassasya86   24    Male      100   
4              4      Iskosia90   23    Male      131   
..           ...            ...  ...     ...      ...   
775          775     Aethedru70   21  Female       60   
776          776         Iral74   21    Male      164   
777          777     Yathecal72   20    Male       67   
778          778        Sisur91    7    Male       92   
779          779      Ennrian78   24    Male       50   

                                     Item Name  Price  
0    Extraction, Quickblade Of Trembling Hands   3.53  
1                            Frenzied Scimitar   1.56  
2                                 Final Critic   4.88  
3                                  Blindscythe   3.27  
4                                         Fury   1.44  
..                                         ...    ...  
775                                       Wolf   3.54  
776                           Exiled Doomblade   1.63  
777      Celeste, Incarnation of the Corrupted   3.46  
778                               Final Critic   4.19  
779                                       Dawn   4.60  

[780 rows x 7 columns]

In [9]:
#check data types
data_file_df.dtypes

Purchase ID      int64
SN              object
Age              int64
Gender          object
Item ID          int64
Item Name       object
Price          float64
dtype: object

In [10]:
#Grabbing the unique values of Items to get Unique Item Count
num_of_items = data_file_df["Item Name"].value_counts()
Number_of_Unique_Items = len(num_of_items)
Number_of_Unique_Items

179

In [11]:
#Sum price to get total Revenue

Total_Revenue = sum(data_file_df["Price"])
                    
Total_Revenue ='${:,.0f}'.format(Total_Revenue)

In [12]:
#get average of the prices through mean function
Average_Price = round(data_file_df["Price"].mean(),2)
Average_Price = '${:.2f}'.format(Average_Price)

In [13]:

Total_Purchase = len(data_file_df)
Total_Purchase ='{:.0f}'.format(Total_Purchase)

In [14]:
summary_df=pd.DataFrame({"Number of Unique Items": [Number_of_Unique_Items],
                         "Average Price": Average_Price,
                         "Total Purchase": [Total_Purchase],
                         "Total Revenue": [Total_Revenue]})
#summary_df[Average_Price] = summary_df[Average_Price].map("${:,.2f}".format)
summary_df

Number of Unique Items Average Price Total Purchase Total Revenue
0                     179         $3.05            780        $2,380

There are 179 Unique items with a total revenue of $2,380 which was purchased 780 times

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [15]:
#drop the duplicates so that you only have one inference to get demographics
Single_value_df= data_file_df.drop_duplicates(subset=['SN'])
#Single_value_df.reset_index()
#value count to summarize m/f/oter, then put in a table(dataframe then format
gender_summary_df = Single_value_df["Gender"].value_counts()
summary_gender_df = pd.DataFrame({"Total Count": gender_summary_df,
                                "% of Players": gender_summary_df/Total_users})
summary_gender_df["% of Players"] = summary_gender_df["% of Players"].map("{:,.2%}".format)

summary_gender_df

Total Count % of Players
Male                           484       84.03%
Female                          81       14.06%
Other / Non-Disclosed           11        1.91%

Observation #1 Large Male Population who purchase products for the game

In [16]:
# SN_df = data_df[["SN","Gender"]].drop_duplicates()
# SN_df["Gender"].value_counts()


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame



Gender	Purchase Count	Average Purchase Price	Total Purchase Value	Avg Total Purchase per Person	
Female	113	$3.20	$361.94	$4.47
Male	652	$3.02	$1,967.64	$4.07
Other / Non-Disclosed	15	$3.35	$50.19	$4.56

In [17]:
data_file_df.columns

Index(['Purchase ID', 'SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [18]:
# Purchase_Count = data_file_df['Gender'].value_counts()
# Purchase_Count
# # Using GroupBy in order to separate the data into fields according to "state" values
# grouped_gender_df = data_file_df.groupby(['Gender'])
# # The object returned is a "GroupBy" object and cannot be viewed normally...
# print(grouped_gender_df)
# # In order to be visualized, a data function must be used...
# Average_Purchase_Price = grouped_gender_df["Price"].mean()
# Average_Purchase_Price
# Total_Purchase_Price = grouped_gender_df["Price"].sum()
# Total_Purchase_Price 
# Avg_Total_Purchase_Price = grouped_gender_df["Price"].sum()/gender_summary_df
# Avg_Total_Purchase_Price

# purchase_analysis_df = pd.DataFrame({"Purchase Count": Purchase_Count,
#                                      "Average Purchase Price":Average_Purchase_Price,
#                                      "Total Purchase Value":Total_Purchase_Price,
#                                      "Avg Total Purchase per Person":Avg_Total_Purchase_Price,
#                                     })
# purchase_analysis_df
    

In [19]:
Purchase_Count = data_file_df['Gender'].value_counts()
Purchase_Count
# Using GroupBy in order to separate the data into fields according to "state" values
grouped_gender_df = data_file_df.groupby(['Gender'])
#Avg_Purchase_Price = grouped_gender_df["Price"].mean()

purchase_analysis_df = pd.DataFrame({"Purchase Count": data_file_df['Gender'].value_counts(),
                                     #"Average Purchase Price": Avg_Purchase_Price,
                                     "Average Purchase Price":grouped_gender_df["Price"].mean().apply(lambda x: f"${x/1:.02f}"),
                                     "Total Purchase Value":grouped_gender_df["Price"].sum().apply(lambda x: f"${x/1:.02f}"),
                                     "Avg Total Purchase per Person":(grouped_gender_df["Price"].sum()/gender_summary_df).apply(lambda x: f"${x/1:.02f}")
                                    })

#Avg_Purchase_Price = '${:.0f}'.format(Avg_Purchase_Price)

purchase_analysis_df
    

Purchase Count Average Purchase Price  \
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

Observation # 2 Though there is a larger Male Population in purchasing items, females spend more dollars on average.  It may be beneficial to target females in purchases to increase frequency and generate higher dollar sales.

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [20]:
#!!!!!!!Sort_values
bins = [0,9,14,19,24,29,34,39,200]
groups_names = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]
Single_value_df["Age Groups"] = pd.cut(Single_value_df["Age"], bins, labels=groups_names, include_lowest=False)
#grouped_age_df = data_file_df.groupby(["Age Groups"])
age_range_sum = Single_value_df["Age Groups"].value_counts()

age_range_sum_df = pd.DataFrame({"Total Count" : age_range_sum,
                                 "% of Players": age_range_sum/Total_users
})

age_range_sum_df["% of Players"] = age_range_sum_df["% of Players"].map("{:,.2%}".format)
age_range_sum_df = age_range_sum_df.sort_index()
age_range_sum_df 

# = age_range_sum_df.sort_values("Age Groups", ascending = False)
# #.sort_values("Age", ascending = False)

/var/folders/fd/z3r1ntv931b76c63_gh69p440000gn/T/ipykernel_28769/2687170705.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  Single_value_df["Age Groups"] = pd.cut(Single_value_df["Age"], bins, labels=groups_names, include_lowest=False)


Total Count % of Players
<10             17        2.95%
10-14           22        3.82%
15-19          107       18.58%
20-24          258       44.79%
25-29           77       13.37%
30-34           52        9.03%
35-39           31        5.38%
40+             12        2.08%

subobservation #1 20-24 year are your core users.  

In [21]:
groups_names = ["<10","10-14","15-19","20-24","25-29","30-34","35-39","40+"]
data_file_df["Age Groups"] = pd.cut(data_file_df["Age"], bins, labels=groups_names, include_lowest=False)
grouped_age_df = data_file_df.groupby(["Age Groups"])

age_range_all_sum = data_file_df["Age Groups"].value_counts()



age_range_sum_all_df = pd.DataFrame({"Total Count" : age_range_all_sum,
                                       "Average Purchase Price":grouped_age_df["Price"].mean().apply(lambda x: f"${x/1:.02f}"),
                                       "Total Purchase Value":grouped_age_df["Price"].sum().apply(lambda x: f"${x/1:.02f}"),
                                      "Avg Total Purchase per Person":(grouped_age_df["Price"].sum()/age_range_sum).apply(lambda x: f"${x/1:.02f}")
                                     })
                                 

# age_range_sum_all_df["% of Players"] = age_range_sum_all_df["% of Players"].map("{:,.2%}".format)

age_range_sum_all_df

Total Count Average Purchase Price Total Purchase Value  \
<10             23                  $3.35               $77.13   
10-14           28                  $2.96               $82.78   
15-19          136                  $3.04              $412.89   
20-24          365                  $3.05             $1114.06   
25-29          101                  $2.90              $293.00   
30-34           73                  $2.93              $214.00   
35-39           41                  $3.60              $147.67   
40+             13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

Observation # 3 16-24 age range is your core demographics, however, your largest basket ring is 35-39 year olds.  Expanding this age group will drive higher dollar sales.

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [22]:
grouped_sn_df = data_file_df.groupby(["SN"])
Top_Spender_df = pd.DataFrame({"Total Count" : grouped_sn_df["SN"].count(),
                              "Average Purchase Price":grouped_sn_df["Price"].mean(),
                              "Total Purchase Value":grouped_sn_df["Price"].sum()})

Top_Spender_df

Top_Spender_df.loc["Lisosia93",]

Sorted_Top_Spender_df = Top_Spender_df.sort_values(["Total Purchase Value"], ascending = False)
Sorted_Top_Spender_df["Total Purchase Value"] = Sorted_Top_Spender_df["Total Purchase Value"].map('${:,.2f}'.format)
Sorted_Top_Spender_df["Average Purchase Price"] =Sorted_Top_Spender_df["Average Purchase Price"].map('${:,.2f}'.format)
Sorted_Top_Spender_df.head()                                
#.apply(lambda x: f"${x:.02f}")  

Total Count Average Purchase Price Total Purchase Value
SN                                                                  
Lisosia93              5                  $3.79               $18.96
Idastidru52            4                  $3.86               $15.45
Chamjask73             3                  $4.61               $13.83
Iral74                 4                  $3.41               $13.62
Iskadarya95            3                  $4.37               $13.10

Observation #4 a promotional event may drive Chamjask73 to spend more as this person spends a higher dollar basket ring than any of the top 5.  

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [23]:
grouped_id_df = data_file_df.groupby(["Item ID","Item Name"])
Most_Popular_df = pd.DataFrame({"Purchase Count" : grouped_id_df["Item ID"].count(),
                              "Item Price":grouped_id_df["Price"].mean().apply(lambda x: f"${x/1:.02f}"),
                              "Total Purchase Value":grouped_id_df["Price"].sum().apply(lambda x: f"${x/1:.02f}")})
Sorted_Most_Popular_df = Most_Popular_df .sort_values(["Purchase Count"], ascending = False)
Sorted_Most_Popular_df.head()        

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [26]:
grouped_id_df = data_file_df.groupby(["Item ID","Item Name"])
Most_Popular_df = pd.DataFrame({"Purchase Count" : grouped_id_df["Item ID"].count(),
                              "Item Price":grouped_id_df["Price"].mean(),
                              "Total Purchase Value":grouped_id_df["Price"].sum()})
Sorted_Most_Popular_df = Most_Popular_df .sort_values(["Total Purchase Value"], ascending = False)
Sorted_Most_Popular_df["Item Price"] = Sorted_Most_Popular_df["Item Price"].map('${:,.2f}'.format)
Sorted_Most_Popular_df["Total Purchase Value"] = Sorted_Most_Popular_df["Total Purchase Value"].map('${:,.2f}'.format)
Sorted_Most_Popular_df.head()        
                                                                                      

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

Observation #5 Nirvana may be a key item to promote on a flash sale, as the item has the largest price and may have room for discount to drive the higher priced item to the top.  Users are willing to pay a higher price for the item.  
Also the purchase frequency needs to be driven higher as the top item only has a purchase count of 13. Bundling and driving flash sales may drive the higher sales of with this group 576 users. There is a large room for improvement in driving item sales. 